# Classifying IMDB movie reviews

In [1]:
import turicreate as tc
import utils

In [2]:
movies = tc.SFrame('./IMDB_Dataset.csv')
movies

Finished parsing file /home/ubuntu/IMDB_Dataset.csv

Parsing completed. Parsed 100 lines in 0.581817 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 39624 lines. Lines per second: 65629.4

Finished parsing file /home/ubuntu/IMDB_Dataset.csv

Parsing completed. Parsed 50000 lines in 0.69352 secs.

Columns:
	review	str
	sentiment	str

Rows: 50000

Data:
+-------------------------------+-----------+
|             review            | sentiment |
+-------------------------------+-----------+
| One of the other reviewers... |  positive |
| A wonderful little product... |  positive |
| I thought this was a wonde... |  positive |
| Basically there's a family... |  negative |
| Petter Mattei's "Love in t... |  positive |
| Probably my all-time favor... |  positive |
| I sure would like to see a... |  positive |
| This show was an amazing, ... |  negative |
| Encouraged by the positive... |  negative |
| If you like original gut w... |  positive |
+-------------------------------+-----------+
[50000 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [3]:
movies['words'] = tc.text_analytics.count_words(movies['review'])
movies

Columns:
	review	str
	sentiment	str
	words	dict

Rows: 50000

Data:
+-------------------------------+-----------+-------------------------------+
|             review            | sentiment |             words             |
+-------------------------------+-----------+-------------------------------+
| One of the other reviewers... |  positive | {'darker': 1.0, 'touch': 1... |
| A wonderful little product... |  positive | {'done': 1.0, 'surface': 1... |
| I thought this was a wonde... |  positive | {'go': 1.0, 'superman': 1.... |
| Basically there's a family... |  negative | {'them': 1.0, 'ignore': 1.... |
| Petter Mattei's "Love in t... |  positive | {'work': 1.0, 'for': 1.0, ... |
| Probably my all-time favor... |  positive | {'for': 1.0, 'dozen': 1.0,... |
| I sure would like to see a... |  positive | {'do': 1.0, 'go': 1.0, 'mu... |
| This show was an amazing, ... |  negative | {'awful': 1.0, 'just': 1.0... |
| Encouraged by the positive... |  negative | {'effort': 1.0, 'an': 1.0,... |
| If you like original gut w... |  positive | {'camp': 1.0, 'great': 1.0... |
+-------------------------------+-----------+-------------------------------+
[50000 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [4]:
model = tc.logistic_classifier.create(movies, features=['words'], target='sentiment')

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 47500

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 101724

Number of coefficients      : 101725

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 2        | 1.000000  | 1.371633     | 0.919284          | 0.864400            |

| 1         | 4        | 1.000000  | 1.821570     | 0.965221          | 0.882800            |

| 2         | 5        | 1.000000  | 2.115625     | 0.975811          | 0.900000            |

| 3         | 7        | 0.927950  | 2.563917     | 0.982653          | 0.904800            |

| 4         | 8        | 1.000000  | 2.851512     | 0.988611          | 0.904800            |

| 9         | 13       | 1.000000  | 4.372810     | 0.999389          | 0.875600            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [5]:
model

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 101725
Number of examples             : 47500
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 101724

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 10
Solver status                  : Completed (Iteration limit reached).
Training time (sec)            : 2.1925

Settings
--------
Log-likelihood                 : 306.8836

Highest Positive Coefficients
-----------------------------
words[queers]                  : 35.8076
words[maced]                   : 35.8076
words[shoring]                 : 35.8076
words[videozone]               : 18.457
words[hrm]                     : 18.2839

Lowest Negative Coefficients
----------------------------
words[undoubtable]             

In [6]:
weights = model.coefficients
weights

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: 101725

Data:
+-------------+---------+----------+------------------------+--------+
|     name    |  index  |  class   |         value          | stderr |
+-------------+---------+----------+------------------------+--------+
| (intercept) |   None  | positive |  0.07567332265252605   |  None  |
|    words    |  darker | positive |   0.6589136198219784   |  None  |
|    words    |  touch  | positive |  0.21798205673836243   |  None  |
|    words    |  thats  | positive |   -0.402304345309385   |  None  |
|    words    |   your  | positive | -0.0037442474446230975 |  None  |
|    words    | viewing | positive |  0.15996992744157226   |  None  |
|    words    |  their  | positive |  0.006185434884117533  |  None  |
|    words    |   into  | positive | -0.018289066947733995  |  None  |
|    words    |  turned | positive |  -0.1909567241039285   |  None  |
|    words    |  being  | positive | -0.005323268752924145  |  None  |
+-------------+---------+----------+------------------------+--------+
[101725 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [7]:
weights.sort('value')

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: 101725

Data:
+-------+----------------+----------+---------------------+--------+
|  name |     index      |  class   |        value        | stderr |
+-------+----------------+----------+---------------------+--------+
| words |  undoubtable   | positive | -28.455286863738227 |  None  |
| words |   cannonite    | positive | -28.455286863738227 |  None  |
| words |     gately     | positive | -16.829449857055202 |  None  |
| words |   millitary    | positive | -16.829449857055202 |  None  |
| words |   nessecary    | positive | -16.829449857055202 |  None  |
| words |    boyzone     | positive | -16.829449857055202 |  None  |
| words |     klines     | positive | -16.829449857055202 |  None  |
| words |   cerimonee    | positive | -16.829449857055202 |  None  |
| words |   bracketts    | positive | -16.829449857055202 |  None  |
| words | unfortuntately | positive | -16.261783026801275 |  None  |
+-------+----------------+----------+---------------------+--------+
[101725 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [8]:
weights.sort('value', ascending=False)

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: 101725

Data:
+-------+------------+----------+--------------------+--------+
|  name |   index    |  class   |       value        | stderr |
+-------+------------+----------+--------------------+--------+
| words |   queers   | positive | 35.80760994044994  |  None  |
| words |   maced    | positive | 35.80760994044994  |  None  |
| words |  shoring   | positive | 35.80760994044994  |  None  |
| words | videozone  | positive | 18.456994202093867 |  None  |
| words |    hrm     | positive | 18.283948686423056 |  None  |
| words |   madcat   | positive | 17.822736499862973 |  None  |
| words |  scribble  | positive | 17.74520459438151  |  None  |
| words | tensionate | positive | 17.580388439632372 |  None  |
| words |   tocsin   | positive | 17.560245658765297 |  None  |
| words |  squirms   | positive | 17.393216774315025 |  None  |
+-------+------------+----------+--------------------+--------+
[101725 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [9]:
weights[weights['index']=='wonderful']

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: Unknown

Data:
+-------+-----------+----------+--------------------+--------+
|  name |   index   |  class   |       value        | stderr |
+-------+-----------+----------+--------------------+--------+
| words | wonderful | positive | 1.0211960859307196 |  None  |
+-------+-----------+----------+--------------------+--------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [10]:
weights[weights['index']=='horrible']

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: Unknown

Data:
+-------+----------+----------+---------------------+--------+
|  name |  index   |  class   |        value        | stderr |
+-------+----------+----------+---------------------+--------+
| words | horrible | positive | -1.0344055288567766 |  None  |
+-------+----------+----------+---------------------+--------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [11]:
weights[weights['index']=='the']

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: Unknown

Data:
+-------+-------+----------+-----------------------+--------+
|  name | index |  class   |         value         | stderr |
+-------+-------+----------+-----------------------+--------+
| words |  the  | positive | 0.0004241444491121867 |  None  |
+-------+-------+----------+-----------------------+--------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [12]:
movies['predictions'] = model.predict(movies, output_type='probability')
movies

Columns:
	review	str
	sentiment	str
	words	dict
	predictions	float

Rows: 50000

Data:
+-------------------------------+-----------+-------------------------------+
|             review            | sentiment |             words             |
+-------------------------------+-----------+-------------------------------+
| One of the other reviewers... |  positive | {'darker': 1.0, 'touch': 1... |
| A wonderful little product... |  positive | {'done': 1.0, 'surface': 1... |
| I thought this was a wonde... |  positive | {'go': 1.0, 'superman': 1.... |
| Basically there's a family... |  negative | {'them': 1.0, 'ignore': 1.... |
| Petter Mattei's "Love in t... |  positive | {'work': 1.0, 'for': 1.0, ... |
| Probably my all-time favor... |  positive | {'for': 1.0, 'dozen': 1.0,... |
| I sure would like to see a... |  positive | {'do': 1.0, 'go': 1.0, 'mu... |
| This show was an amazing, ... |  negative | {'awful': 1.0, 'just': 1.0... |
| Encouraged by the positive... |  negative | {'effort': 1.0, 'an': 1.0,... |
| If you like original gut w... |  positive | {'camp': 1.0, 'great': 1.0... |
+-------------------------------+-----------+-------------------------------+
+-----------------------+
|      predictions      |
+-----------------------+
|   0.9999994640410024  |
|   0.999999999831477   |
|   0.9929302718333678  |
|  0.02304713916264185  |
|   0.9985597321712851  |
|   0.9995842353595462  |
|   0.9999011506692725  |
|  0.008012263599900784 |
| 5.531662102661318e-07 |
|   0.9674411283578057  |
+-----------------------+
[50000 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [28]:
# returns the movie object with the highest "most positive" value for the "predictions" attribute.
movies.sort('predictions')[-1]

{'review': 'Jackie Chan\'s Police Story is a landmark film for both the Honk Kong action genre and the career of Jackie Chan.<br /><br />Directed/written by Chan, Police Story has a basic plot as did all the films of that era and genre, and like most of the the films of Police Storys\' kind, the script is nothing to be raved about. But the plot of the film is Jackie Chan, who plays a nice guy cop, struggling to convict the local gang lord.<br /><br />The direction of the film is nothing special and by no means the best directing effort that Jackie Chan has given us, that responsibility falls to the underrated masterpiece "Miracles". However the job that Jackie does directing is sufficient and respectable. The standout out directing of the film comes with the fight scenes.<br /><br />The performances in this film also vary with Jackie giving a very solid typical Chan nice guy up against it role, but this is by no means his best acting role, that can been seen in the Sammo Hung directed 

In [33]:
movies.sort('predictions')[-1]

{'review': 'Jackie Chan\'s Police Story is a landmark film for both the Honk Kong action genre and the career of Jackie Chan.<br /><br />Directed/written by Chan, Police Story has a basic plot as did all the films of that era and genre, and like most of the the films of Police Storys\' kind, the script is nothing to be raved about. But the plot of the film is Jackie Chan, who plays a nice guy cop, struggling to convict the local gang lord.<br /><br />The direction of the film is nothing special and by no means the best directing effort that Jackie Chan has given us, that responsibility falls to the underrated masterpiece "Miracles". However the job that Jackie does directing is sufficient and respectable. The standout out directing of the film comes with the fight scenes.<br /><br />The performances in this film also vary with Jackie giving a very solid typical Chan nice guy up against it role, but this is by no means his best acting role, that can been seen in the Sammo Hung directed 

In [29]:
# Most negative review
movies.sort('predictions')[0]

{'review': "Even duller, if possible, than the original (I hope I may say that under the IMDb guidelines). THE FRENCH CONNECTION at least tried to absorb European influences, to complicate the conventional view of the American police detective, even if the attempt was foundered by Friedkin's ambivalence, Americaness and general indirection. The (relative) arthouse boom of the 1960s (especially with the nouvelle vague) allowed for a huge influence of European cinema in Hollywood. This lent a new vigour and complexity to a weary medium, and, in the best of them (eg BONNIE AND CLYDE, early Scorcese), a new subversion of received practice. The original CONNECTION was part of this movement, with its difficultly distanced style, and anti-detective figure. TWO is old Hollywood's right-wing reassertion of American values.<br /><br />This is figured in the film's very tiresome America vs France dialectic. For instance, TWO is shot like a 70s French policier. It was the French, of course, who in

In [27]:
movies.sort('predictions', ascending=True)[0]

{'review': "Even duller, if possible, than the original (I hope I may say that under the IMDb guidelines). THE FRENCH CONNECTION at least tried to absorb European influences, to complicate the conventional view of the American police detective, even if the attempt was foundered by Friedkin's ambivalence, Americaness and general indirection. The (relative) arthouse boom of the 1960s (especially with the nouvelle vague) allowed for a huge influence of European cinema in Hollywood. This lent a new vigour and complexity to a weary medium, and, in the best of them (eg BONNIE AND CLYDE, early Scorcese), a new subversion of received practice. The original CONNECTION was part of this movement, with its difficultly distanced style, and anti-detective figure. TWO is old Hollywood's right-wing reassertion of American values.<br /><br />This is figured in the film's very tiresome America vs France dialectic. For instance, TWO is shot like a 70s French policier. It was the French, of course, who in

### Assignement 3 Starts here: 

### 1) Use the Sentiment_Analysis_IMDB.ipynb notebook to answer the following questions. Insert text cells to write out your answers. Be sure to state the question number.

#### a) [2 marks] How many items are in the raw dataset? What data is contained in each item, and what is the label?

In [34]:
movies.shape

(50000, 4)

There were 50000 items, each one has the comment review, as a string, in the first column and and the tag of positive or negative sentiment in the second column. 

### b) [2 marks] What does the tc.text_analytics.count_words() function do?

This command line transform the comment review (string) into a dictionary with the number of repetition of each word as am imdependient feature.

### c) [2 marks] How much of the data is used for validation? What are the training and validation accuracies, and what does this imply in terms of overfitting/underfitting?

The data is divided in Training and validation with the proportion 0.95 to 0.5, the default values of splitting data. 

Training acurracy: 0.999389          | Validating Acurracy: 0.875600           

The model is overfitting, the testing acurracy is better than validation acurracy.  


### d) [1 mark] How many coefficients are in the model?


In [38]:
# Get the number of coefficients in the model
num_coefficients = len(model.coefficients)

print("Number of coefficients in the logistic regression model:", num_coefficients)

Number of coefficients in the logistic regression model: 101725



### e) [1 mark] What weights are given to the L1 and L2 penalties for regularization?


In [44]:
# Get the L1 and L2 penalty values of the model
l1_penalty = model.l1_penalty
l2_penalty = model.l2_penalty

print("L1 penalty:", l1_penalty)
print("L2 penalty:", l2_penalty)

L1 penalty: 0.0
L2 penalty: 0.01



### f) [2 marks] Compare the top 10 words that contributed to positive reviews with the top 10 words that contributed to negative reviews. What is the difference between them?


In [50]:
negative = weights.sort('value').head()
negative


Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: 10

Data:
+-------+----------------+----------+---------------------+--------+
|  name |     index      |  class   |        value        | stderr |
+-------+----------------+----------+---------------------+--------+
| words |  undoubtable   | positive | -28.455286863738227 |  None  |
| words |   cannonite    | positive | -28.455286863738227 |  None  |
| words |     gately     | positive | -16.829449857055202 |  None  |
| words |   millitary    | positive | -16.829449857055202 |  None  |
| words |   nessecary    | positive | -16.829449857055202 |  None  |
| words |    boyzone     | positive | -16.829449857055202 |  None  |
| words |     klines     | positive | -16.829449857055202 |  None  |
| words |   cerimonee    | positive | -16.829449857055202 |  None  |
| words |   bracketts    | positive | -16.829449857055202 |  None  |
| words | unfortuntately | positive | -16.261783026801275 |  None  |
+-------+----------------+----------+---------------------+--------+
[10 rows x 5 columns]

In [49]:
positive = weights.sort('value', ascending=False).head()
positive

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: 10

Data:
+-------+------------+----------+--------------------+--------+
|  name |   index    |  class   |       value        | stderr |
+-------+------------+----------+--------------------+--------+
| words |   queers   | positive | 35.80760994044994  |  None  |
| words |   maced    | positive | 35.80760994044994  |  None  |
| words |  shoring   | positive | 35.80760994044994  |  None  |
| words | videozone  | positive | 18.456994202093867 |  None  |
| words |    hrm     | positive | 18.283948686423056 |  None  |
| words |   madcat   | positive | 17.822736499862973 |  None  |
| words |  scribble  | positive | 17.74520459438151  |  None  |
| words | tensionate | positive | 17.580388439632372 |  None  |
| words |   tocsin   | positive | 17.560245658765297 |  None  |
| words |  squirms   | positive | 17.393216774315025 |  None  |
+-------+------------+----------+--------------------+--------+
[10 rows x 5 columns]

Positive values are higher, 
Both are classified as positive in the original Tag


### g) [1 mark] What weight is given to the word ‘the’? Is this a sensible value? Why or why not?


In [62]:
the = weights[weights['index']=='the']
print('the value is: ', the['value'])

the value is:  [0.0004241444491121867, ... ]


The value of the is positive but too low to influence the final result, so it could be considered as neutral


### h) [2 marks] Describe what you would do to improve this model, and why it would help.

Suggestions: 
1) Data cleaning and preprocessing: The provided dataset may contain irrelevant or noisy information that could hinder model performance. Cleaning and preprocessing the dataset can improve model accuracy.

2) Feature engineering: Instead of using a simple bag-of-words representation, we can extract more informative features such as n-grams, word embeddings, or topic models. This can help the model capture more nuanced relationships between words and improve accuracy.

3) Model selection and hyperparameter tuning: We can experiment with different machine learning models such as Support Vector Machines, Random Forests, or Neural Networks to see if they perform better than logistic regression. We can also tune hyperparameters such as the regularization strength or learning rate to improve model performance.

4) Evaluating performance on a test set: The provided code trains and evaluates the model on the same dataset. However, it is essential to evaluate the model on a separate test set to estimate its generalization performance accurately. For instance, Train/Validate/Test in the proportion 80/10/10 %. This can help identify issues such as overfitting or underfitting and allow us to fine-tune the model accordingly.

5) Ensemble learning: We can combine multiple models to improve overall performance. This can be done through techniques such as bagging, boosting, or stacking. Changing the hyperparameters in the configuration of the model, L1, L2, spochs, model. 

6) Training at least two model with different hyperparameters and make a comparison between.
